In [207]:
import pandas as pd
import numpy as np

In [3]:
matches2010 = pd.read_csv('atp_matches_2010.csv')

In [67]:
matches2010[:-19]

tourney_id   tourney_name surface  draw_size tourney_level  tourney_date  \
0      2010-339       Brisbane    Hard         32             A      20100103   
1      2010-339       Brisbane    Hard         32             A      20100103   
2      2010-339       Brisbane    Hard         32             A      20100103   
3      2010-339       Brisbane    Hard         32             A      20100103   
4      2010-339       Brisbane    Hard         32             A      20100103   
...         ...            ...     ...        ...           ...           ...   
3006   2010-352  Paris Masters    Hard         48             M      20101107   
3007   2010-352  Paris Masters    Hard         48             M      20101107   
3008   2010-605    Tour Finals    Hard          8             F      20101121   
3009   2010-605    Tour Finals    Hard          8             F      20101121   
3010   2010-605    Tour Finals    Hard          8             F      20101121   

      match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0             1     104053          1.0          NaN  ...    34.0     29.0   
1             2     104958          NaN           WC  ...    34.0     22.0   
2             3     104755          NaN          NaN  ...    58.0     38.0   
3             4     105051          NaN            Q  ...    29.0     16.0   
4             5     104607          4.0          NaN  ...    41.0     26.0   
...         ...        ...          ...          ...  ...     ...      ...   
3006         46     104417          4.0          NaN  ...    65.0     57.0   
3007         47     104417          4.0          NaN  ...    30.0     21.0   
3008          1     103819          NaN          NaN  ...    39.0     23.0   
3009          2     103819          NaN          NaN  ...    27.0     16.0   
3010          3     103819          NaN          NaN  ...    39.0     31.0   

      l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0         11.0    10.0        3.0        5.0          7.0             4410.0   
1         14.0     9.0        7.0       10.0        134.0              400.0   
2         14.0    14.0        7.0       11.0         52.0              850.0   
3         15.0     9.0        2.0        5.0        285.0              151.0   
4         14.0     9.0        6.0        9.0         20.0             1655.0   
...        ...     ...        ...        ...          ...                ...   
3006      22.0    18.0        2.0        4.0          5.0             4960.0   
3007      12.0     9.0        1.0        3.0          5.0             4960.0   
3008      13.0     8.0        5.0        9.0          2.0             7645.0   
3009      19.0     9.0        5.0        8.0          2.0             7645.0   
3010      15.0    10.0        2.0        4.0          2.0             7645.0   

     loser_rank loser_rank_points  
0          77.0             598.0  
1          78.0             590.0  
2          88.0             568.0  
3          28.0            1260.0  
4         251.0             179.0  
...         ...               ...  
3006       34.0            1215.0  
3007       14.0            2485.0  
3008        7.0            3735.0  
3009        5.0            5360.0  
3010        4.0            5380.0  

[3011 rows x 49 columns]

In [5]:
matches2010.columns

Index(['tourney_id', 'tourney_name', 'surface', 'draw_size', 'tourney_level',
       'tourney_date', 'match_num', 'winner_id', 'winner_seed', 'winner_entry',
       'winner_name', 'winner_hand', 'winner_ht', 'winner_ioc', 'winner_age',
       'loser_id', 'loser_seed', 'loser_entry', 'loser_name', 'loser_hand',
       'loser_ht', 'loser_ioc', 'loser_age', 'score', 'best_of', 'round',
       'minutes', 'w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon',
       'w_SvGms', 'w_bpSaved', 'w_bpFaced', 'l_ace', 'l_df', 'l_svpt',
       'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced',
       'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points'],
      dtype='object')

In [8]:
'''

[ser_ht', 'loser_ioc', 'loser_age', 'score', 'best_of', 'round',
       'minutes', 'w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon',
       'w_SvGms', 'w_bpSaved', 'w_bpFaced', 'l_ace', 'l_df', 'l_svpt',
       'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced',
       'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points'],
      dtype='object')'tourney_id', 'tourney_name', 'surface', 'draw_size', 'tourney_level',
       'tourney_date', 'match_num', 'winner_id', 'winner_seed', 'winner_entry',
       'winner_name', 'winner_hand', 'winner_ht', 'winner_ioc', 'winner_age',
       'loser_id', 'loser_seed', 'loser_entry', 'loser_name', 'loser_hand',
       'lo

    
    ['tourney_name', 'surface', 'winner_id', 
       'winner_name', 'winner_hand', 'winner_ht', 'winner_age',
       'loser_id', 'loser_name', 'loser_hand',
       'loser_ht', 'loser_age', 'score', 'best_of']
      dtype='object')

'''

"\n\n['tourney_id', 'tourney_name', 'surface', 'draw_size', 'tourney_level',\n       'tourney_date', 'match_num', 'winner_id', 'winner_seed', 'winner_entry',\n       'winner_name', 'winner_hand', 'winner_ht', 'winner_ioc', 'winner_age',\n       'loser_id', 'loser_seed', 'loser_entry', 'loser_name', 'loser_hand',\n       'loser_ht', 'loser_ioc', 'loser_age', 'score', 'best_of', 'round',\n       'minutes', 'w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon',\n       'w_SvGms', 'w_bpSaved', 'w_bpFaced', 'l_ace', 'l_df', 'l_svpt',\n       'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced',\n       'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points'],\n      dtype='object')\n\n    \n    ['tourney_name', 'surface', 'winner_id', \n       'winner_name', 'winner_hand', 'winner_ht', 'winner_age',\n       'loser_id', 'loser_name', 'loser_hand',\n       'loser_ht', 'loser_age', 'score', 'best_of']\n      dtype='object')\n\n"

In [31]:
new2010 = matches2010[['tourney_name', 'surface', 'winner_id', 
       'winner_name', 'winner_hand', 'winner_ht', 'winner_age',
       'loser_id', 'loser_name', 'loser_hand',
       'loser_ht', 'loser_age', 'best_of', 'winner_rank', 'loser_rank']]

In [28]:
len(new2010)

3030

In [54]:
new2010.iloc[3000]

tourney_name      Paris Masters
surface                    Hard
tourney_level                 M
winner_id                103566
winner_name      Michael Llodra
winner_hand                   L
winner_ht                 190.0
winner_age            30.475017
loser_id                 104925
loser_name       Novak Djokovic
loser_hand                    R
loser_ht                  188.0
loser_age             23.466119
best_of                       3
winner_rank                34.0
loser_rank                  3.0
Name: 3000, dtype: object

In [36]:
## loser rank is lower than winner rank, aka what we would expect
new2010.iloc[299][-1] < new2010.iloc[200][-2]

True

In [52]:
new2010.iloc[0][8:14], new2010.iloc[0][-1]

(loser_id            103429
 loser_name    Peter Luczak
 loser_hand               R
 loser_ht             183.0
 loser_age        30.346338
 best_of                  3
 Name: 0, dtype: object,
 77.0)

In [55]:
# def countwinnerslosers(df):
#     winners = []
#     losers = []
#     for i in range len(df)-1:
#         winners.append(df.iloc[i][1:7], df.iloc[i][-2])
#         losers.append(df.iloc[i][8:14], df.iloc[i][])
        

In [68]:
courts = pd.get_dummies(new2010['surface'])

In [70]:
courts # nice

Carpet  Clay  Grass  Hard
0          0     0      0     1
1          0     0      0     1
2          0     0      0     1
3          0     0      0     1
4          0     0      0     1
...      ...   ...    ...   ...
3025       0     0      0     1
3026       0     0      0     1
3027       0     1      0     0
3028       0     1      0     0
3029       0     1      0     0

[3030 rows x 4 columns]

In [71]:
winnerhand = pd.get_dummies(new2010['winner_hand'])
loserhand = pd.get_dummies(new2010['loser_hand'])

In [74]:
type(winnerhand)

pandas.core.frame.DataFrame

In [75]:
new2010

tourney_name surface tourney_level  winner_id  \
0                       Brisbane    Hard             A     104053   
1                       Brisbane    Hard             A     104958   
2                       Brisbane    Hard             A     104755   
3                       Brisbane    Hard             A     105051   
4                       Brisbane    Hard             A     104607   
...                          ...     ...           ...        ...   
3025  Davis Cup WG F: SRB vs FRA    Hard             D     104925   
3026  Davis Cup WG F: SRB vs FRA    Hard             D     104678   
3027                  Dusseldorf    Clay             A     104724   
3028                  Dusseldorf    Clay             A     103908   
3029                  Dusseldorf    Clay             A     104676   

             winner_name winner_hand  winner_ht  winner_age  loser_id  \
0           Andy Roddick           R      188.0   27.348392    103429   
1           Carsten Ball           L      198.0   22.543463    104999   
2        Richard Gasquet           R      185.0   23.548255    103813   
3          Matthew Ebden           R      188.0   22.108145    103781   
4          Tomas Berdych           R      196.0   24.298426    105194   
...                  ...         ...        ...         ...       ...   
3025      Novak Djokovic           R      188.0   23.523614    104792   
3026      Viktor Troicki           R      193.0   24.799452    103566   
3027     Eduardo Schwank           R      183.0   24.065708    103908   
3028  Paul Henri Mathieu           R      185.0   28.342231    105583   
3029        Andreas Beck           L      190.0   24.276523    103908   

              loser_name loser_hand  loser_ht  loser_age  best_of  \
0           Peter Luczak          R     183.0  30.346338        3   
1          Mischa Zverev          L     190.0  22.370979        3   
2        Jarkko Nieminen          L     185.0  28.451745        3   
3          Jurgen Melzer          L     183.0  28.621492        3   
4           Nick Lindahl          R     183.0  21.429158        3   
...                  ...        ...       ...        ...      ...   
3025        Gael Monfils          R     193.0  24.243669        5   
3026      Michael Llodra          L     190.0  30.532512        5   
3027  Paul Henri Mathieu          R     185.0  28.342231        3   
3028       Dusan Lajovic          R     180.0  19.879535        3   
3029  Paul Henri Mathieu          R     185.0  28.342231        3   

      winner_rank  loser_rank  
0             7.0        77.0  
1           134.0        78.0  
2            52.0        88.0  
3           285.0        28.0  
4            20.0       251.0  
...           ...         ...  
3025          3.0        12.0  
3026         30.0        23.0  
3027         53.0        56.0  
3028         56.0       504.0  
3029         80.0        56.0  

[3030 rows x 16 columns]

In [85]:
carpets = courts.iloc[:,0]

In [88]:
carpets

0       0
1       0
2       0
3       0
4       0
       ..
3025    0
3026    0
3027    0
3028    0
3029    0
Name: Carpet, Length: 3030, dtype: uint8

In [89]:
new2010.insert(16,'carpet', carpets)

In [91]:
new2010 #fucking works. choice

tourney_name surface tourney_level  winner_id  \
0                       Brisbane    Hard             A     104053   
1                       Brisbane    Hard             A     104958   
2                       Brisbane    Hard             A     104755   
3                       Brisbane    Hard             A     105051   
4                       Brisbane    Hard             A     104607   
...                          ...     ...           ...        ...   
3025  Davis Cup WG F: SRB vs FRA    Hard             D     104925   
3026  Davis Cup WG F: SRB vs FRA    Hard             D     104678   
3027                  Dusseldorf    Clay             A     104724   
3028                  Dusseldorf    Clay             A     103908   
3029                  Dusseldorf    Clay             A     104676   

             winner_name winner_hand  winner_ht  winner_age  loser_id  \
0           Andy Roddick           R      188.0   27.348392    103429   
1           Carsten Ball           L      198.0   22.543463    104999   
2        Richard Gasquet           R      185.0   23.548255    103813   
3          Matthew Ebden           R      188.0   22.108145    103781   
4          Tomas Berdych           R      196.0   24.298426    105194   
...                  ...         ...        ...         ...       ...   
3025      Novak Djokovic           R      188.0   23.523614    104792   
3026      Viktor Troicki           R      193.0   24.799452    103566   
3027     Eduardo Schwank           R      183.0   24.065708    103908   
3028  Paul Henri Mathieu           R      185.0   28.342231    105583   
3029        Andreas Beck           L      190.0   24.276523    103908   

              loser_name loser_hand  loser_ht  loser_age  best_of  \
0           Peter Luczak          R     183.0  30.346338        3   
1          Mischa Zverev          L     190.0  22.370979        3   
2        Jarkko Nieminen          L     185.0  28.451745        3   
3          Jurgen Melzer          L     183.0  28.621492        3   
4           Nick Lindahl          R     183.0  21.429158        3   
...                  ...        ...       ...        ...      ...   
3025        Gael Monfils          R     193.0  24.243669        5   
3026      Michael Llodra          L     190.0  30.532512        5   
3027  Paul Henri Mathieu          R     185.0  28.342231        3   
3028       Dusan Lajovic          R     180.0  19.879535        3   
3029  Paul Henri Mathieu          R     185.0  28.342231        3   

      winner_rank  loser_rank  carpet  
0             7.0        77.0       0  
1           134.0        78.0       0  
2            52.0        88.0       0  
3           285.0        28.0       0  
4            20.0       251.0       0  
...           ...         ...     ...  
3025          3.0        12.0       0  
3026         30.0        23.0       0  
3027         53.0        56.0       0  
3028         56.0       504.0       0  
3029         80.0        56.0       0  

[3030 rows x 17 columns]

In [92]:
v2 = pd.read_csv('atp_matches_2010.csv')

In [95]:
cols = list(v2.columns)

In [98]:
start = v2[['surface','tourney_level','best_of','winner_hand','winner_ht','winner_age','winner_rank','loser_hand','loser_ht','loser_age','loser_rank']]

In [99]:
start

surface tourney_level  best_of winner_hand  winner_ht  winner_age  \
0       Hard             A        3           R      188.0   27.348392   
1       Hard             A        3           L      198.0   22.543463   
2       Hard             A        3           R      185.0   23.548255   
3       Hard             A        3           R      188.0   22.108145   
4       Hard             A        3           R      196.0   24.298426   
...      ...           ...      ...         ...        ...         ...   
3025    Hard             D        5           R      188.0   23.523614   
3026    Hard             D        5           R      193.0   24.799452   
3027    Clay             A        3           R      183.0   24.065708   
3028    Clay             A        3           R      185.0   28.342231   
3029    Clay             A        3           L      190.0   24.276523   

      winner_rank loser_hand  loser_ht  loser_age  loser_rank  
0             7.0          R     183.0  30.346338        77.0  
1           134.0          L     190.0  22.370979        78.0  
2            52.0          L     185.0  28.451745        88.0  
3           285.0          L     183.0  28.621492        28.0  
4            20.0          R     183.0  21.429158       251.0  
...           ...        ...       ...        ...         ...  
3025          3.0          R     193.0  24.243669        12.0  
3026         30.0          L     190.0  30.532512        23.0  
3027         53.0          R     185.0  28.342231        56.0  
3028         56.0          R     180.0  19.879535       504.0  
3029         80.0          R     185.0  28.342231        56.0  

[3030 rows x 11 columns]

In [105]:
courts = pd.get_dummies(start['surface'])
carpet = courts.iloc[:,0]
clay = courts.iloc[:,1]
grass = courts.iloc[:,2]
hard = courts.iloc[:,3]
## now have the court data split and one-hotted, can DROP court data and re-insert these series, methinks

In [108]:
start.drop('surface', axis=1, inplace=True)

/Users/MBest/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4315: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [111]:
start.insert(0, 'carpet', carpet)

ValueError: cannot insert carpet, already exists

In [114]:

start.insert(2, 'clay', clay)
start.insert(3, 'grass', grass)
start.insert(4, 'hard', hard)

In [115]:
start

carpet tourney_level  clay  grass  hard  best_of winner_hand  winner_ht  \
0          0             A     0      0     1        3           R      188.0   
1          0             A     0      0     1        3           L      198.0   
2          0             A     0      0     1        3           R      185.0   
3          0             A     0      0     1        3           R      188.0   
4          0             A     0      0     1        3           R      196.0   
...      ...           ...   ...    ...   ...      ...         ...        ...   
3025       0             D     0      0     1        5           R      188.0   
3026       0             D     0      0     1        5           R      193.0   
3027       0             A     1      0     0        3           R      183.0   
3028       0             A     1      0     0        3           R      185.0   
3029       0             A     1      0     0        3           L      190.0   

      winner_age  winner_rank loser_hand  loser_ht  loser_age  loser_rank  
0      27.348392          7.0          R     183.0  30.346338        77.0  
1      22.543463        134.0          L     190.0  22.370979        78.0  
2      23.548255         52.0          L     185.0  28.451745        88.0  
3      22.108145        285.0          L     183.0  28.621492        28.0  
4      24.298426         20.0          R     183.0  21.429158       251.0  
...          ...          ...        ...       ...        ...         ...  
3025   23.523614          3.0          R     193.0  24.243669        12.0  
3026   24.799452         30.0          L     190.0  30.532512        23.0  
3027   24.065708         53.0          R     185.0  28.342231        56.0  
3028   28.342231         56.0          R     180.0  19.879535       504.0  
3029   24.276523         80.0          R     185.0  28.342231        56.0  

[3030 rows x 14 columns]

In [116]:
tourneys = pd.get_dummies(start['tourney_level'])

In [118]:
tourneys
A = tourneys.iloc[:,0]
D = tourneys.iloc[:,1]
F = tourneys.iloc[:,2]
G = tourneys.iloc[:,3]
M = tourneys.iloc[:,4]

In [122]:
start.drop('tourney_level', axis=1, inplace=True)


/Users/MBest/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4315: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [132]:
winhand = pd.get_dummies(start['winner_hand'])


In [133]:
L = winhand.iloc[:,0]
R = winhand.iloc[:,1]
U = winhand.iloc[:,2]

In [135]:
start.drop('winner_hand', axis=1, inplace=True)

/Users/MBest/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4315: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [137]:
start.insert(8, '1L', L)

In [139]:
start.insert(9, '1R', R)

In [141]:
start.insert(10, '1U', U)

In [142]:
start

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
0          0     0      0     1        3      188.0   27.348392          7.0   
1          0     0      0     1        3      198.0   22.543463        134.0   
2          0     0      0     1        3      185.0   23.548255         52.0   
3          0     0      0     1        3      188.0   22.108145        285.0   
4          0     0      0     1        3      196.0   24.298426         20.0   
...      ...   ...    ...   ...      ...        ...         ...          ...   
3025       0     0      0     1        5      188.0   23.523614          3.0   
3026       0     0      0     1        5      193.0   24.799452         30.0   
3027       0     1      0     0        3      183.0   24.065708         53.0   
3028       0     1      0     0        3      185.0   28.342231         56.0   
3029       0     1      0     0        3      190.0   24.276523         80.0   

      1L  1R  1U loser_hand  loser_ht  loser_age  loser_rank  
0      0   1   0          R     183.0  30.346338        77.0  
1      1   0   0          L     190.0  22.370979        78.0  
2      0   1   0          L     185.0  28.451745        88.0  
3      0   1   0          L     183.0  28.621492        28.0  
4      0   1   0          R     183.0  21.429158       251.0  
...   ..  ..  ..        ...       ...        ...         ...  
3025   0   1   0          R     193.0  24.243669        12.0  
3026   0   1   0          L     190.0  30.532512        23.0  
3027   0   1   0          R     185.0  28.342231        56.0  
3028   0   1   0          R     180.0  19.879535       504.0  
3029   1   0   0          R     185.0  28.342231        56.0  

[3030 rows x 15 columns]

In [143]:
losehands = pd.get_dummies(start['loser_hand'])

In [148]:
L2 = losehands.iloc[:,0]
R2 = losehands.iloc[:,1]
U2 = losehands.iloc[:,2]

In [145]:
start.drop('loser_hand', axis=1, inplace=True)

/Users/MBest/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4315: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [149]:
start.insert(14, 'L2', L2)
start.insert(15, 'R2', R2)
start.insert(16, 'U2', U2)

In [172]:
start

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
0          0     0      0     1        3      188.0   27.348392          7.0   
1          0     0      0     1        3      198.0   22.543463        134.0   
2          0     0      0     1        3      185.0   23.548255         52.0   
3          0     0      0     1        3      188.0   22.108145        285.0   
4          0     0      0     1        3      196.0   24.298426         20.0   
...      ...   ...    ...   ...      ...        ...         ...          ...   
3025       0     0      0     1        5      188.0   23.523614          3.0   
3026       0     0      0     1        5      193.0   24.799452         30.0   
3027       0     1      0     0        3      183.0   24.065708         53.0   
3028       0     1      0     0        3      185.0   28.342231         56.0   
3029       0     1      0     0        3      190.0   24.276523         80.0   

      1L  1R  1U  loser_ht  loser_age  loser_rank  L2  R2  U2  
0      0   1   0     183.0  30.346338        77.0   0   1   0  
1      1   0   0     190.0  22.370979        78.0   1   0   0  
2      0   1   0     185.0  28.451745        88.0   1   0   0  
3      0   1   0     183.0  28.621492        28.0   1   0   0  
4      0   1   0     183.0  21.429158       251.0   0   1   0  
...   ..  ..  ..       ...        ...         ...  ..  ..  ..  
3025   0   1   0     193.0  24.243669        12.0   0   1   0  
3026   0   1   0     190.0  30.532512        23.0   1   0   0  
3027   0   1   0     185.0  28.342231        56.0   0   1   0  
3028   0   1   0     180.0  19.879535       504.0   0   1   0  
3029   1   0   0     185.0  28.342231        56.0   0   1   0  

[3030 rows x 17 columns]

In [156]:
def jumblewinners(df):
    evens = []
    for index, val in df.iterrows():
        if index % 2 == 0:
            evens.append(val)
            
    return evens

In [162]:
jumblewinners(start)[0][:5] ##match attributes

carpet     0.0
clay       0.0
grass      0.0
hard       1.0
best_of    3.0
Name: 0, dtype: float64

In [166]:
jumblewinners(start)[0][5:11]  ##player1 attributes

winner_ht      188.000000
winner_age      27.348392
winner_rank      7.000000
1L               0.000000
1R               1.000000
1U               0.000000
Name: 0, dtype: float64

In [171]:
jumblewinners(start)[0][-6:]  # player2 attributes... can we switch them???? what about removing/replacing them?
# I cant think




loser_ht      183.000000
loser_age      30.346338
loser_rank     77.000000
L2              0.000000
R2              1.000000
U2              0.000000
Name: 0, dtype: float64

In [177]:
b2010 = start ## This is a working set of 3030 matches, i want to flip half the winners and losers and add winner data as well...



In [178]:
evens2010 = b2010[b2010.index % 2 == 0] # keep these the same for now

In [185]:
odds2010 = b2010[b2010.index % 2 != 0]


In [203]:
odds2010


carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
1          0     0      0     1        3      198.0   22.543463        134.0   
3          0     0      0     1        3      188.0   22.108145        285.0   
5          0     0      0     1        3      183.0   24.550308         42.0   
7          0     0      0     1        3      188.0   22.015058         36.0   
9          0     0      0     1        3      188.0   32.766598         58.0   
...      ...   ...    ...   ...      ...        ...         ...          ...   
3021       0     0      0     1        3      188.0   23.504449          3.0   
3023       0     0      0     1        5      193.0   24.243669         12.0   
3025       0     0      0     1        5      188.0   23.523614          3.0   
3027       0     1      0     0        3      183.0   24.065708         53.0   
3029       0     1      0     0        3      190.0   24.276523         80.0   

      1L  1R  1U  loser_ht  loser_age  loser_rank  L2  R2  U2  
1      1   0   0     190.0  22.370979        78.0   1   0   0  
3      0   1   0     183.0  28.621492        28.0   1   0   0  
5      0   1   0     188.0  28.071184        55.0   0   1   0  
7      1   0   0     190.0  30.349076        73.0   0   1   0  
9      0   1   0     185.0  25.782341        80.0   0   1   0  
...   ..  ..  ..       ...        ...         ...  ..  ..  ..  
3021   0   1   0     188.0  28.229979         8.0   0   1   0  
3023   0   1   0     180.0  26.436687        49.0   0   1   0  
3025   0   1   0     193.0  24.243669        12.0   0   1   0  
3027   0   1   0     185.0  28.342231        56.0   0   1   0  
3029   1   0   0     185.0  28.342231        56.0   0   1   0  

[1515 rows x 17 columns]

In [194]:
# def switchwinnerloser(oddsdf):
#     loserinfo = []
#     winnerinfo = []
#     for i, val in oddsdf.iterrows():
#         loserinfo.append(oddsdf.iloc[i][-6:])
#         winnerinfo.append(oddsdf.iloc[i][5:11])
        
#     return loserinfo, winnerinfo

In [201]:
switchedodds = odds2010[['carpet', 'clay', 'grass', 'hard', 'best_of', 'loser_ht', 'loser_age', 'loser_rank',
       'L2', 'R2', 'U2', 'winner_ht', 'winner_age','winner_rank', '1L', '1R', '1U']]

In [204]:
switchedodds ## almost too simple? 

carpet  clay  grass  hard  best_of  loser_ht  loser_age  loser_rank  L2  \
1          0     0      0     1        3     190.0  22.370979        78.0   1   
3          0     0      0     1        3     183.0  28.621492        28.0   1   
5          0     0      0     1        3     188.0  28.071184        55.0   0   
7          0     0      0     1        3     190.0  30.349076        73.0   0   
9          0     0      0     1        3     185.0  25.782341        80.0   0   
...      ...   ...    ...   ...      ...       ...        ...         ...  ..   
3021       0     0      0     1        3     188.0  28.229979         8.0   0   
3023       0     0      0     1        5     180.0  26.436687        49.0   0   
3025       0     0      0     1        5     193.0  24.243669        12.0   0   
3027       0     1      0     0        3     185.0  28.342231        56.0   0   
3029       0     1      0     0        3     185.0  28.342231        56.0   0   

      R2  U2  winner_ht  winner_age  winner_rank  1L  1R  1U  
1      0   0      198.0   22.543463        134.0   1   0   0  
3      0   0      188.0   22.108145        285.0   0   1   0  
5      1   0      183.0   24.550308         42.0   0   1   0  
7      1   0      188.0   22.015058         36.0   1   0   0  
9      1   0      188.0   32.766598         58.0   0   1   0  
...   ..  ..        ...         ...          ...  ..  ..  ..  
3021   1   0      188.0   23.504449          3.0   0   1   0  
3023   1   0      193.0   24.243669         12.0   0   1   0  
3025   1   0      188.0   23.523614          3.0   0   1   0  
3027   1   0      183.0   24.065708         53.0   0   1   0  
3029   1   0      190.0   24.276523         80.0   1   0   0  

[1515 rows x 17 columns]

In [224]:
from numpy import random
reversedorder = ['carpet', 'clay', 'grass', 'hard', 'best_of', 'loser_ht', 'loser_age','loser_rank','L2', 'R2', 'U2', 'winner_ht', 'winner_age','winner_rank', '1L', '1R', '1U']

In [247]:
keys = reorders.values()

In [274]:
reorders = {
    'winner_ht': 'a_ht',
    'winner_age': 'a_age',
    'winner_rank': 'a_rank',
    '1L': 'aL',
    '1R': 'aR',
    '1U': 'aU',
    'loser_ht': 'b_ht',
    'loser_age': 'b_age',
    'loser_rank': 'b_rank',
    'L2': 'bL',
    'R2': 'bR',
    'U2': 'bU'
    }



def flipornot(row):

    flip = np.random.choice([0,1], p=[0.5,0.5])
    
    if flip == 0: #player info does not shuffle, player A is winner (default)
        row['result'] = 0
    else:   #winner info goes into player B spot
        row = row.reindex(reversedorder)
        row['result'] = 1
    #rename here
    
    return row.rename(dict(zip(row.index[-len(keys):], keys)))


viewnew = b2010.apply(lambda x: flipornot(x), axis=1)

In [230]:
b2010

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
0          0     0      0     1        3      188.0   27.348392          7.0   
1          0     0      0     1        3      198.0   22.543463        134.0   
2          0     0      0     1        3      185.0   23.548255         52.0   
3          0     0      0     1        3      188.0   22.108145        285.0   
4          0     0      0     1        3      196.0   24.298426         20.0   
...      ...   ...    ...   ...      ...        ...         ...          ...   
3025       0     0      0     1        5      188.0   23.523614          3.0   
3026       0     0      0     1        5      193.0   24.799452         30.0   
3027       0     1      0     0        3      183.0   24.065708         53.0   
3028       0     1      0     0        3      185.0   28.342231         56.0   
3029       0     1      0     0        3      190.0   24.276523         80.0   

      1L  1R  1U  loser_ht  loser_age  loser_rank  L2  R2  U2  
0      0   1   0     183.0  30.346338        77.0   0   1   0  
1      1   0   0     190.0  22.370979        78.0   1   0   0  
2      0   1   0     185.0  28.451745        88.0   1   0   0  
3      0   1   0     183.0  28.621492        28.0   1   0   0  
4      0   1   0     183.0  21.429158       251.0   0   1   0  
...   ..  ..  ..       ...        ...         ...  ..  ..  ..  
3025   0   1   0     193.0  24.243669        12.0   0   1   0  
3026   0   1   0     190.0  30.532512        23.0   1   0   0  
3027   0   1   0     185.0  28.342231        56.0   0   1   0  
3028   0   1   0     180.0  19.879535       504.0   0   1   0  
3029   1   0   0     185.0  28.342231        56.0   0   1   0  

[3030 rows x 17 columns]

In [232]:
np.random.choice([0,1], p=[0.5,0.5])

1

In [275]:
viewnew

aL   aR     aU  a_age       a_ht  a_rank   bL   bR   bU  b_age  \
0     1.0  0.0  183.0    7.0  27.348392     0.0  1.0  0.0  0.0   77.0   
1     0.0  0.0  198.0   78.0  22.370979     1.0  0.0  0.0  1.0  134.0   
2     1.0  0.0  185.0   52.0  23.548255     0.0  0.0  0.0  0.0   88.0   
3     0.0  0.0  188.0   28.0  28.621492     1.0  1.0  0.0  1.0  285.0   
4     1.0  0.0  183.0   20.0  24.298426     0.0  1.0  0.0  0.0  251.0   
...   ...  ...    ...    ...        ...     ...  ...  ...  ...    ...   
3025  1.0  0.0  193.0    3.0  23.523614     0.0  1.0  0.0  0.0   12.0   
3026  0.0  0.0  193.0   23.0  30.532512     1.0  1.0  0.0  1.0   30.0   
3027  1.0  0.0  185.0   53.0  24.065708     0.0  1.0  0.0  0.0   56.0   
3028  1.0  0.0  185.0  504.0  19.879535     0.0  1.0  0.0  1.0   56.0   
3029  0.0  0.0  185.0   80.0  24.276523     1.0  1.0  0.0  0.0   56.0   

           b_ht  b_rank  best_of  carpet  clay  grass  hard  loser_ht  \
0     30.346338     0.0      3.0     0.0   0.0    0.0   1.0       NaN   
1     22.543463     1.0      3.0     0.0   0.0    0.0   1.0     190.0   
2     28.451745     1.0      3.0     0.0   0.0    0.0   1.0       NaN   
3     22.108145     0.0      3.0     0.0   0.0    0.0   1.0     183.0   
4     21.429158     0.0      3.0     0.0   0.0    0.0   1.0       NaN   
...         ...     ...      ...     ...   ...    ...   ...       ...   
3025  24.243669     0.0      5.0     0.0   0.0    0.0   1.0       NaN   
3026  24.799452     0.0      5.0     0.0   0.0    0.0   1.0     190.0   
3027  28.342231     0.0      3.0     0.0   1.0    0.0   0.0       NaN   
3028  28.342231     0.0      3.0     0.0   1.0    0.0   0.0     180.0   
3029  28.342231     0.0      3.0     0.0   1.0    0.0   0.0       NaN   

      winner_ht  
0         188.0  
1           NaN  
2         185.0  
3           NaN  
4         196.0  
...         ...  
3025      188.0  
3026        NaN  
3027      183.0  
3028        NaN  
3029      190.0  

[3030 rows x 19 columns]

In [281]:
reorders = {
    'winner_ht': 'a_ht',
    'winner_age': 'a_age',
    'winner_rank': 'a_rank',
    '1L': 'aL',
    '1R': 'aR',
    '1U': 'aU',
    'loser_ht': 'b_ht',
    'loser_age': 'b_age',
    'loser_rank': 'b_rank',
    'L2': 'bL',
    'R2': 'bR',
    'U2': 'bU'
    }

matchatt = {
    'carpet': 'carpet', 
    'clay': 'clay',
    'grass': 'grass', 
    'hard': 'hard', 
    'best_of': 'best_of'
}



def flipornotnewframe(row):
    newframe = []
    flip = np.random.choice([0,1], p=[0.5,0.5])
    
    if flip == 0: #player info does not shuffle, player A is winner (default)
        row['result'] = 0
        newframe.append(row)
        
    else:   #winner info goes into player B spot
        row = row.reindex(reversedorder)
        row['result'] = 1
        newframe.append(row)
        
    #rename here
    #holder = dict(zip(row.index[-len(keys):], keys))
    #return row.rename(matchatt.update(holder))
    return pd.DataFrame(newframe)

try2 = b2010.apply(lambda x: flipornotnewframe(x), axis=1)

In [302]:
try2[6] #this actually kind of works, the labels are still 'attached' but it doesn't matter, does it?

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
6     0.0   0.0    0.0   1.0      3.0      185.0   31.416838        119.0   

    1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  result  
6  0.0  1.0  0.0     190.0  29.631759        67.0  1.0  0.0  0.0     0.0

In [330]:
try2.loc[:-1]

Series([], dtype: object)

In [342]:
#session working?
"lets try2"

'lets try2'

In [343]:
try2

0          carpet  clay  grass  hard  best_of  winner_...
1          carpet  clay  grass  hard  best_of  winner_...
2          carpet  clay  grass  hard  best_of  loser_h...
3          carpet  clay  grass  hard  best_of  winner_...
4          carpet  clay  grass  hard  best_of  winner_...
                              ...                        
3025          carpet  clay  grass  hard  best_of  winn...
3026          carpet  clay  grass  hard  best_of  winn...
3027          carpet  clay  grass  hard  best_of  lose...
3028          carpet  clay  grass  hard  best_of  lose...
3029          carpet  clay  grass  hard  best_of  winn...
Length: 3030, dtype: object

In [347]:
try2[3] #certainly seems to be working, luckily .

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
3     0.0   0.0    0.0   1.0      3.0      188.0   22.108145        285.0   

    1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  result  
3  0.0  1.0  0.0     183.0  28.621492        28.0  1.0  0.0  0.0     0.0

In [349]:
# def geth2hforplayer(matches,name):
#     """get all head-to-heads of the player"""
#     matches = matches[(matches['winner_name'] == name) | (matches['loser_name'] == name)]
#     h2hs = {}
#     for index, match in matches.iterrows():
#         if (match['winner_name'] == name):
#             if (match['loser_name'] not in h2hs):
#                 h2hs[match['loser_name']] = {}
#                 h2hs[match['loser_name']]['l'] = 0
#                 h2hs[match['loser_name']]['w'] = 1
#             else:
#                 h2hs[match['loser_name']]['w'] = h2hs[match['loser_name']]['w']+1
#         elif (match['loser_name'] == name):
#             if (match['winner_name'] not in h2hs):
#                 h2hs[match['winner_name']] = {}
#                 h2hs[match['winner_name']]['w'] = 0
#                 h2hs[match['winner_name']]['l'] = 1
#             else:
#                 h2hs[match['winner_name']]['l'] = h2hs[match['winner_name']]['l']+1

#Random Forest Attempt

In [499]:
from random import randrange, random
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, precision_score, recall_score, accuracy_score, plot_confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
from sklearn import tree

In [367]:
mytest = try2[1]

In [401]:
mytest

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
1     0.0   0.0    0.0   1.0      3.0      198.0   22.543463        134.0   

    1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  result  
1  1.0  0.0  0.0     190.0  22.370979        78.0  1.0  0.0  0.0     0.0

In [382]:
mytest.iloc[:,:-1]

carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
1     0.0   0.0    0.0   1.0      3.0      198.0   22.543463        134.0   

    1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  
1  1.0  0.0  0.0     190.0  22.370979        78.0  1.0  0.0  0.0

In [379]:
mytest.result


1    0.0
Name: result, dtype: float64

In [385]:
def makeitworkplease(df):
    X = []
    y = []
    for i, row in df.iterrows():
        X.append(df.iloc[:-1])
        y.append(df.result)
        
    return X, y
        

In [386]:
makeitworkplease(try2)

AttributeError: 'Series' object has no attribute 'iterrows'

In [387]:
thirdtry = try2.apply(lambda x: x)

In [388]:
thirdtry

0          carpet  clay  grass  hard  best_of  winner_...
1          carpet  clay  grass  hard  best_of  winner_...
2          carpet  clay  grass  hard  best_of  loser_h...
3          carpet  clay  grass  hard  best_of  winner_...
4          carpet  clay  grass  hard  best_of  winner_...
                              ...                        
3025          carpet  clay  grass  hard  best_of  winn...
3026          carpet  clay  grass  hard  best_of  winn...
3027          carpet  clay  grass  hard  best_of  lose...
3028          carpet  clay  grass  hard  best_of  lose...
3029          carpet  clay  grass  hard  best_of  winn...
Length: 3030, dtype: object

KeyError: -12

In [435]:
noresults[0][1]


carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
1     0.0   0.0    0.0   1.0      3.0      198.0   22.543463        134.0   

    1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  result  
1  1.0  0.0  0.0     190.0  22.370979        78.0  1.0  0.0  0.0     0.0

In [436]:
#ok fuck this, im just going to copy the dataframe, drop the results, and pull them another way or some shit.

In [463]:
try3 = try2.copy()

In [464]:
type(try3)

pandas.core.series.Series

In [466]:
def delresult(srs):
    srs = srs.iloc[-1]
    return srs

In [485]:
please = try3.apply(lambda x: delresult(x))

In [5]:
type(please)

NameError: name 'please' is not defined

In [487]:
working = please.drop('result', axis=1)

In [489]:
working #fucking finally


carpet  clay  grass  hard  best_of  winner_ht  winner_age  winner_rank  \
0        0.0   0.0    0.0   1.0      3.0      188.0   27.348392          7.0   
1        0.0   0.0    0.0   1.0      3.0      198.0   22.543463        134.0   
2        0.0   0.0    0.0   1.0      3.0      185.0   23.548255         52.0   
3        0.0   0.0    0.0   1.0      3.0      188.0   22.108145        285.0   
4        0.0   0.0    0.0   1.0      3.0      196.0   24.298426         20.0   
...      ...   ...    ...   ...      ...        ...         ...          ...   
3025     0.0   0.0    0.0   1.0      5.0      188.0   23.523614          3.0   
3026     0.0   0.0    0.0   1.0      5.0      193.0   24.799452         30.0   
3027     0.0   1.0    0.0   0.0      3.0      183.0   24.065708         53.0   
3028     0.0   1.0    0.0   0.0      3.0      185.0   28.342231         56.0   
3029     0.0   1.0    0.0   0.0      3.0      190.0   24.276523         80.0   

       1L   1R   1U  loser_ht  loser_age  loser_rank   L2   R2   U2  
0     0.0  1.0  0.0     183.0  30.346338        77.0  0.0  1.0  0.0  
1     1.0  0.0  0.0     190.0  22.370979        78.0  1.0  0.0  0.0  
2     0.0  1.0  0.0     185.0  28.451745        88.0  1.0  0.0  0.0  
3     0.0  1.0  0.0     183.0  28.621492        28.0  1.0  0.0  0.0  
4     0.0  1.0  0.0     183.0  21.429158       251.0  0.0  1.0  0.0  
...   ...  ...  ...       ...        ...         ...  ...  ...  ...  
3025  0.0  1.0  0.0     193.0  24.243669        12.0  0.0  1.0  0.0  
3026  0.0  1.0  0.0     190.0  30.532512        23.0  1.0  0.0  0.0  
3027  0.0  1.0  0.0     185.0  28.342231        56.0  0.0  1.0  0.0  
3028  0.0  1.0  0.0     180.0  19.879535       504.0  0.0  1.0  0.0  
3029  1.0  0.0  0.0     185.0  28.342231        56.0  0.0  1.0  0.0  

[3030 rows x 17 columns]

In [514]:
X = working
y = please.result

In [520]:
X.isnull().values.any()
y.isnull().values.any() ## my Random forest would not work yesterday due to Nans, today I rerun and no nan. K.

False

In [508]:
y # we made it.

In [509]:
## I need to understand WHY that is so fucking annoying because I can't do that 30 more times

In [510]:
##begin random forest## 

In [518]:
#train test split to enable model evaluation
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3)

In [519]:
# Instantiate model with 1000 decision trees
rf = RandomForestClassifier(n_estimators = 1000)
# Train the model on training data
rf.fit(X_train, y_train)


RandomForestClassifier(n_estimators=1000)

In [513]:
# I WOULD LIKE TO SMASH YOU INTO PIECES, COMPUTER #

In [521]:
#ok, we're actually going to start doing machine learning 7/12

In [522]:
predictions = rf.predict(X_test)

In [547]:
errors = abs(predictions - y_test)
type(errors)
errors.isnull().values.any()
y_test.isnull().values.any()

False

In [556]:
np.mean(errors)
np.mean(y_test)

errors - y_test
y_test / errors

383     1.0
1442    1.0
727     inf
2065    inf
2389    0.0
       ... 
522     1.0
2419    1.0
1257    0.0
2571    inf
2881    0.0
Name: result, Length: 909, dtype: float64

In [525]:
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

Mean Absolute Error: 0.47 degrees.


In [577]:
# Calculate mean absolute percentage error (MAPE)
# mape = 100 * (errors / che)
# mape

383     100.0
1442    100.0
727       0.0
2065      0.0
2389      inf
        ...  
522     100.0
2419    100.0
1257      inf
2571      0.0
2881      inf
Name: result, Length: 909, dtype: float64

In [597]:
def checkformatch(a,b):
    correct = 0
    for i, v in enumerate(zip(a,b)):
        if a[i] == b[i]:
            correct += 1
        else:
            pass
    
    return correct, len(a), correct/len(a)

In [576]:
type(predictions)
len(predictions)

909

i had to cast y_test to an array for some reason

In [573]:
checkans = np.array(y_test)

In [574]:
type(checkans)

numpy.ndarray

In [575]:
checkformatch(predictions, checkans)

485

485 correct out of 909 = 0.53355 accuracy (or precision? or recall? lol) for attempt 1

In [580]:
## incredibly ugly, lets try again 


In [581]:
#im feeling dumb so lets just increase the tree size and see if we learn more this time#

In [583]:
#train test split to enable model evaluation
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3)

In [584]:
# Instantiate model with 10000 decision trees
rf = RandomForestClassifier(n_estimators = 10000)   ##class_weight will be very nice to explain age, court, etc mean more than l/r/u perhaps
# Train the model on training data
rf.fit(X_train, y_train)


RandomForestClassifier(n_estimators=10000)

In [585]:
predictions2 = rf.predict(X_test)

In [589]:
len(predictions2)

909

In [598]:
checkformatch(predictions2, np.array(y_test))

(456, 909, 0.5016501650165016)

In [590]:
## 456 correct out of 909...literally worse. 

In [599]:
#how would I do if I randomly guess 909 ouctomes??

In [606]:
def randomguess(num):
    guess = []
    for i in range(num):
        flip = np.random.choice([0,1], p=[0.5,0.5])
        guess.append(flip)
    
    return np.array(guess)

In [609]:
p= randomguess(909)

In [610]:
checkformatch(p, np.array(y_test))

(474, 909, 0.5214521452145214)

In [612]:
# my random guesses did better than random forest this time 


In [613]:
#will entropy be helpful?

In [614]:
#train test split to enable model evaluation
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3)

In [615]:
# Instantiate model with 1000 decision trees
rf = RandomForestClassifier(n_estimators = 1000, criterion='entropy' )   ##class_weight will be very nice to explain age, court, etc mean more than l/r/u perhaps
# Train the model on training data
rf.fit(X_train, y_train)


RandomForestClassifier(criterion='entropy', n_estimators=1000)

In [619]:
predictions3 = rf.predict(X_test)

In [620]:
checkformatch(predictions3, np.array(y_test))

(460, 909, 0.506050605060506)

In [621]:
from pprint import pprint


In [622]:
pprint(rf.get_params())

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'entropy',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 1000,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}


In [623]:
from sklearn.model_selection import RandomizedSearchCV


In [624]:
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
pprint(random_grid)
{'bootstrap': [True, False],
 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}

{'bootstrap': [True, False],
 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}


{'bootstrap': [True, False],
 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}

In [625]:
# Use the random grid to search for best hyperparameters
# First create the base model to tune
rf = RandomForestClassifier()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_random = RandomizedSearchCV(estimator = rf, param_distributions = random_grid, n_iter = 100, cv = 3, verbose=2, random_state=42, n_jobs = -1)
# Fit the random search model
rf_random.fit(X_train, y_train)

Fitting 3 folds for each of 100 candidates, totalling 300 fits


RandomizedSearchCV(cv=3, estimator=RandomForestClassifier(), n_iter=100,
                   n_jobs=-1,
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': [10, 20, 30, 40, 50, 60,
                                                      70, 80, 90, 100, 110,
                                                      None],
                                        'max_features': ['auto', 'sqrt'],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': [200, 400, 600, 800,
                                                         1000, 1200, 1400, 1600,
                                                         1800, 2000]},
                   random_state=42, verbose=2)

In [626]:
rf_random.best_params_

{'n_estimators': 200,
 'min_samples_split': 2,
 'min_samples_leaf': 1,
 'max_features': 'sqrt',
 'max_depth': 50,
 'bootstrap': True}

In [627]:
# {'n_estimators': 200,
#  'min_samples_split': 2,
#  'min_samples_leaf': 1,
#  'max_features': 'sqrt',
#  'max_depth': 50,
#  'bootstrap': True}

In [628]:
def evaluate(model, test_features, test_labels):
    predictions = model.predict(test_features)
    errors = abs(predictions - test_labels)
    mape = 100 * np.mean(errors / test_labels)
    accuracy = 100 - mape
    print('Average Error: {:0.4f} degrees.'.format(np.mean(errors)))
    print('Accuracy = {:0.2f}%.'.format(accuracy))
    
    return accuracy

In [632]:
rf = RandomForestClassifier(n_estimators=200, min_samples_split=2, min_samples_leaf=1, max_features='sqrt', max_depth=50, bootstrap=True)
##this has all the 'best' random hyperparams

In [633]:
rf.fit(X_train, y_train)

RandomForestClassifier(max_depth=50, max_features='sqrt', n_estimators=200)

In [635]:
tuned = rf.predict(X_test)

In [638]:
checkformatch(tuned, np.array(y_test)) ## not any better. great.

(461, 909, 0.5071507150715071)

In [3]:
## ok, we have a "working" model that performs meh, 2 routes to go:

## 1 - simpler way: add more data - lets clean 5 years of matches and see if we can improve significantly

## 2 - add more info TO the data - grab aces/match by year (or service games one, etc, some extra metrics, 
##but will be needed to be added for both players X 3000 matches per year, sounds brutal


In [4]:
games2011 = pd.read_csv('atp_matches_2011.csv')

NameError: name 'pd' is not defined